### PR TITLE
Add optimization tutorial

### PR DESCRIPTION
**Pull request by gaurav-arya**

### CODE REVIEW

File: docs/src/tutorials/optimizations.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+lines!(ax, ps, avg, label = "≈ E[X(p)]")
+lines!(ax, ps, derivative, label = "≈ d/dp E[X(p)]")
+vlines!(ax, [p_opt], label = "p_opt", color = :green, linewidth = 2.0)
+hlines!(ax, [0.0], color = :black, linewidth = 1.0)
+
+f[1, 2] = Legend(f, ax, framevisible = false)
+ylims!(ax, (-50, 80))
+ax = f[2, 1:2] = Axis(f, title = "Optimizer trace", xlabel="Iterations", ylabel="Value of p")
+lines!(ax, trace, color = :green, linewidth = 2.0)
+save("crazy_opt.png", f,  px_per_unit = 4)
```
Comment (frankschae) [inline docs/src/tutorials/optimizations.md:69]:
```suggestion
save("crazy_opt.png", f,  px_per_unit = 4) # hide
```

File: docs/src/tutorials/optimizations.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+lines!(ax, ps, avg, label = "≈ E[X(p)]")
+lines!(ax, ps, derivative, label = "≈ d/dp E[X(p)]")
+vlines!(ax, [p_opt], label = "p_opt", color = :green, linewidth = 2.0)
+hlines!(ax, [0.0], color = :black, linewidth = 1.0)
+
+f[1, 2] = Legend(f, ax, framevisible = false)
+ylims!(ax, (-10, 10))
+ax = f[2, 1:2] = Axis(f, title = "Optimizer trace", ylabel="Value of p", xlabel="Iterations")
+lines!(ax, trace, color = :green, linewidth = 2.0)
+save("variational.png", f, px_per_unit = 4)
```
Comment (frankschae) [inline docs/src/tutorials/optimizations.md:117]:
```suggestion 
save("variational.png", f, px_per_unit = 4) # hide
```

File: docs/src/tutorials/optimizations.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
+# Stochastic optimization with discreteness
+
+```@setup random_walk
+import Pkg
+Pkg.activate("../../../tutorials/toy_optimizations")
+Pkg.develop(path="../../..")
+Pkg.instantiate()
+```
+
+In this tutorial, we solve two stochastic optimization problems using `StochasticAD` where the objective contains discrete randomness. We will need the following packages:
```
Comment (frankschae) [inline docs/src/tutorials/optimizations.md:10]: "the objective contains" sounds a bit imprecise.

File: docs/src/tutorials/optimizations.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+Let's maximize $\mathbb{E}[X(p)]$! First, let's setup the problem:
+```@example optimizations
+p0 = [0.5] # initial value of p
+m = StochasticAD.StochasticModel(p0, x -> -X(x)) # Formulate as minimization problem
+```
+Now, let's perform stochastic gradient descent using [Adam](https://arxiv.org/abs/1412.6980).
+```@example optimizations
+# Minimize E[X] = KL(Poisson(p)| NegativeBinomial(10, 1-30/(10+30)))
```
Comment (frankschae) [inline docs/src/tutorials/optimizations.md:37]: This seems to refer to the other example.

**Comment by mschauer** [comment]
Doc preview https://gaurav-arya.github.io/StochasticAD.jl/previews/PR31/

**Comment by mschauer** [comment]
The NegativeBinomial approximation of a Poisson from here https://en.wikipedia.org/wiki/Negative_binomial_distribution#Poisson_distribution

File: docs/src/tutorials/optimizations.md (hunk 3 above)
Comment (mschauer) [reply docs/src/tutorials/optimizations.md:10]: "involves samples from discrete distributions"?

File: docs/src/tutorials/optimizations.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+```@example optimizations
+function X(p)
+    a = p * (1 - p)
+    b = rand(Binomial(10, p))
+    c = 2 * b + 3 * rand(Bernoulli(p))
+    return a * c * rand(Normal(b, a))
+end
+```
+
+Let's maximize $\mathbb{E}[X(p)]$! First, let's setup the problem, using the [`StochasticModel`](@ref)helper utility to create a trainable model:
```
Comment (mschauer) [inline docs/src/tutorials/optimizations.md:30]:
```suggestion
Let's maximize $\mathbb{E}[X(p)]$! First, let's setup the problem, using the [`StochasticModel`](@ref) helper utility to create a trainable model:
```

File: docs/src/tutorials/optimizations.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+```@example optimizations
+p0 = [0.5] # initial value of p
+m = StochasticAD.StochasticModel(p0, x -> -X(x)) # Formulate as minimization problem
```
Comment (mschauer) [inline docs/src/tutorials/optimizations.md:33]:
```suggestion
m = StochasticAD.StochasticModel(p0, x -> -X(x)) # formulate as minimization problem
```

File: docs/src/tutorials/optimizations.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+ylims!(ax, (-50, 80))
+ax = f[2, 1:2] = Axis(f, title = "Optimizer trace", xlabel="Iterations", ylabel="Value of p")
+lines!(ax, trace, color = :green, linewidth = 2.0)
+save("crazy_opt.png", f,  px_per_unit = 4) # hide
+```
+![](crazy_opt.png)
+
+## Solving a variational problem
+
+Let's consider a toy variational program: we find a[Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution) that is close to that of a [negative Binomial](https://en.wikipedia.org/wiki/Negative_binomial_distribution), via minimization of the [Kullback-Leibler divergence](https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence) $D_{\mathrm{KL}}$. Concretly, let us solve
```
Comment (mschauer) [inline docs/src/tutorials/optimizations.md:75]:
```suggestion
Let's consider a toy variational program: we find a [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution) that is close to that of a [negative Binomial](https://en.wikipedia.org/wiki/Negative_binomial_distribution), via minimization of the [Kullback-Leibler divergence](https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence) $D_{\mathrm{KL}}$. Concretly, let us solve
```

File: docs/src/tutorials/optimizations.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+```math
+\underset{p \in \mathbb{R}}{\operatorname{argmin}} D_{\mathrm{KL}}\left(\mathrm{Pois}(p) \middle\| \mathrm{NB}(10, 0.25) \right).
```
Comment (mschauer) [inline docs/src/tutorials/optimizations.md:77]:
```suggestion
\underset{p \in \mathbb{R}}{\operatorname{argmin}}\; D_{\mathrm{KL}}\left(\mathrm{Pois}(p) \middle\| \mathrm{NB}(10, 0.25) \right).
```

File: docs/src/tutorials/optimizations.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+```
+The following program produces an unbiased estimate of the objective:
+```@example optimizations
+function X(p)
+    i = rand(Poisson(p))
+    return logpdf(Poisson(p), i) - logpdf(NegativeBinomial(10, 1 - 30 / (10 + 30)), i)
+end
+```
+We can now optimize the KL divergence via stochastic gradient descent!
```
Comment (mschauer) [inline docs/src/tutorials/optimizations.md:86]:
```suggestion
We can now optimize the KL-divergence via stochastic gradient descent!
```

**Comment by mschauer** [comment]
Last thing to check tomorrow: Consistency in choosing direction of KL

File: docs/src/tutorials/optimizations.md [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+Let's consider a toy variational program: we find a[Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution) that is close to that of a [negative Binomial](https://en.wikipedia.org/wiki/Negative_binomial_distribution), via minimization of the [Kullback-Leibler divergence](https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence) $D_{\mathrm{KL}}$. Concretly, let us solve
+```math
+\underset{p \in \mathbb{R}}{\operatorname{argmin}} D_{\mathrm{KL}}\left(\mathrm{Pois}(p) \middle\| \mathrm{NB}(10, 0.25) \right).
+```
+The following program produces an unbiased estimate of the objective:
+```@example optimizations
+function X(p)
+    i = rand(Poisson(p))
+    return logpdf(Poisson(p), i) - logpdf(NegativeBinomial(10, 1 - 30 / (10 + 30)), i)
```
Comment (mschauer) [inline docs/src/tutorials/optimizations.md:83]:
```suggestion
    return logpdf(Poisson(p), i) - logpdf(NegativeBinomial(10, 0.25), i)
```

**Review by mschauer — APPROVED** [review]

**Review by frankschae — APPROVED** [review]

File: docs/src/public_api.md [hunk 11]
```diff
@@ -14,5 +14,17 @@ StochasticAD.perturbations
...
-What happens if we run [`derivative_contribution`](@ref) after each step, instead of only at the end? This is *smoothing*, which combines the second and third components of a single stochastic triple into a single dual component. Smoothing no longer has a guarantee of unbiasedness, but is surprisingly accurate in a number of situations. [*Smoothing functionality coming soon.*]
+## Smoothing
 
+What happens if we run [`derivative_contribution`](@ref) after each step, instead of only at the end? This is *smoothing*, which combines the second and third components of a single stochastic triple into a single dual component. Smoothing no longer has a guarantee of unbiasedness, but is surprisingly accurate in a number of situations. 
+
+[*Smoothing functionality coming soon.*]
+
+## Optimization
+
+We also provide a couple utilities to make it easier to get started with forming and training a model via stochastic gradient descent:
```
Comment (mschauer) [inline docs/src/public_api.md:25]:
```suggestion
We also provide utilities to make it easier to get started with forming and training a model via stochastic gradient descent:
```
